### PR TITLE
[UIE-172] Package useLoadedData hook

### PR DESCRIPTION
**Pull request by nawatts**
`useLoadedData` and its variants are widely used hooks that will need to be package in a low-level package so that they can be used by multiple high level packages. This moves them to the components package.

This also required packaging a few dependencies: `usePrevious` and `useStore`. `useStore` is re-exported from react-utils to avoid updating a bunch of import paths.

### CODE REVIEW

File: packages/components/src/hooks/useAutoLoadedData.test.ts [hunk 1]
```diff
@@ -1,9 +1,9 @@
 import { LoadedState } from '@terra-ui-packages/core-utils';
 import { act, renderHook } from '@testing-library/react';
-import { UseLoadedDataResult } from 'src/libs/ajax/loaded-data/useLoadedData';
-import { controlledPromise } from 'src/testing/test-utils';
```
Comment (cahrens) [inline packages/components/src/hooks/useAutoLoadedData.test.ts:4]: What about the `controlledPromise` in `src/testing/test-utils`?

Comment (nawatts) [reply packages/components/src/hooks/useAutoLoadedData.test.ts:4]: To try to limit the changes in this PR, I copied it into components' test-utils module.

For a follow up, I intent to move it to core-utils and have both the components package and the app use it from there.

Comment (cahrens) [reply packages/components/src/hooks/useAutoLoadedData.test.ts:4]: OK, I wanted to make sure the duplication wasn't permanent.

File: packages/components/src/hooks/useStore.test.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    const { result: hookReturnRef, rerender } = renderHook((args: [Atom<string>]) => useSettableStore(...args), {
+      initialProps: [myAtom],
+    });
+    const [initialValue, setValue] = hookReturnRef.current;
+
+    // Act
+    act(() => {
+      setValue('goodbye');
+    });
+    rerender([myAtom]);
```
Comment (cahrens) [inline packages/components/src/hooks/useStore.test.ts]: Why do you need to rerender? The documentation says that the component will rerender when the store changes.

Comment (nawatts) [reply packages/components/src/hooks/useStore.test.ts]: This was an existing test and I didn't look at it too closely when I copied it. But yeah, this looks like it invalidates the test.

Comment (nawatts) [reply packages/components/src/hooks/useStore.test.ts]: Fixed in https://github.com/DataBiosphere/terra-ui/pull/4805/commits/da23b4c556f50e70f5d754ab0f28aed7207f431c.

**Review by cahrens — APPROVED** [review]

**Review by pshapiro4broad — APPROVED** [review]
Looks good to me

**Review by LizBaldo — APPROVED** [review]

File: packages/components/src/hooks/useAutoLoadedData.test.ts (hunk 1 above)
Comment (nawatts) [reply packages/components/src/hooks/useAutoLoadedData.test.ts:4]: Done in https://github.com/DataBiosphere/terra-ui/pull/4810.

**Pull request merged**